### ISSUE TITLE
Issues when doing releases on `rel-X.Y` branches

### ISSUE THREAD

**Issue by laurenceisla** [PostgREST/postgrest#3113]
Was going to comment on #3082, but it's out of the scope of that issue.

### Problem

When doing a release from a `rel-X.Y` branch, then the `bump to vA.B` commit is only present in the `rel-A.B` branch. That's why we have to manually create a CHANGELOG commit in `main`, but not bump the `postgrest.cabal` version. This causes 2 things:
  - The next patch release needs to cherry pick the commits (although this is OK for back patches)
  - Makes the release process more painful. If the main branch has all the commits for the release, then it should be more straightforward to bump the version.
  - The `postgrest.cabal` version in `main` is permanently outdated (right now it's `11.2.0` when `12.0.1` is the latest release)

### Proposal

Before the release is done, verify if the `rel-A.B` and `main` branches are pointing to the same commit ( the same `git rev-parse <branch>`), if so, then the bump commit can be merged into `main` and pushed too (pushes `rel-A.B`, `main` and `vA.B.C`). This could be done automatically or ask if it's OK to merge.

If the branches are not pointing to the same commit, then automatically add only the CHANGELOG update to main, or ask if it's OK, or just inform to do it (I almost forgot to do that for `v12.0.1`).

**Comment by steve-chavez** [comment]
Q: The CHANGELOG can be the same on the main branch and the rel branch right? Doesn't matter if we have some unreleased additions on main and if we do a patch version. After all users will only see the main branch CHANGELOG. 

---

Another issue with the CHANGELOG is that it's no longer good for major or minor releases, because I have to edit it manually to link the features to the docs as on https://github.com/PostgREST/postgrest/releases/tag/v12.0.0

---

Maybe we should no longer update the CHANGELOG manually on PRs (this is easy to forget too). Whenever a `postgrest-release` is done, only then the CHANGELOG should be updated automatically. Then this commit can go on the main branch and on the release branch.

For this we can stick to a commit convention (similar to the one Wolfgang suggested before), but that also includes a link to the latest docs.

Not sure about the cabal version, I think for now we can just duplicate it. Although this would matter more if https://github.com/PostgREST/postgrest/issues/2275 was done. 

Perhaps the correct thing for main would be to have the highest version plus the date as the cabal version, like we did on the nightlies. So right now it'd be 12.0.1.20231215.

**Comment by laurenceisla** [comment]
> I think this could work pretty well.

@wolfgangwalther Agree, I like the idea. That way main would be always 1 step ahead of the latest release instead of perpetually stuck in an older version.

> No dates anywhere in version number anymore. Those just cause trouble, because they always compare "bigger" than regular version numbers, even if in fact older.

+1 they gave more headaches than anything.

**Comment by steve-chavez** [comment]
> We want an updated pre-release after each commit on main to the github release page and also push a docker image each time.

> After each commit to main, we make a pre-relase and docker release with <major>.<minor> in postgrest.cabal (so 12.1 or 12.3 etc.), but on github releases and dockerhub we tag this as <major>.<minor>-nightly. The nightly suffix will prevent dependabot and renovate from updating to/from this - they will always want to keep the suffix or avoid adding it. We are going to replace the existing pre-release / nightly tag.

Hm, will we do that for pure docs commits too? (it doesn't seem right) That should happen once https://github.com/PostgREST/postgrest/pull/3120 is merged.

Maybe we should finish #3120 before settling on a new release model. It will simplify a lot of things.

**Comment by steve-chavez** [comment]
> We already do docs releases on each commit. The latest tag on the docs is the version from the postgrest-docs main branch. That should stay the same in the future. We just need to also make nightly releases on every commit, so that docs and latest nightly are actually in sync.

Right, forgot about that - docs do get a release for each commit on main.

> Maybe we can think about only having one version of the docs online for each major version? The latest minor release of that, ofc.

Agree.

> Taking the "publish docs on each commit on release branches" idea one step further: I think we should do the same for bugfixes.
> Ultimately, we can get a bot to work for backpatching

That sounds perfect/ideal.. though I don't know the effort it takes. It seems considerable with the bot and all.

> Compare that with [renovate](https://github.com/renovatebot/renovate/releases), which truly releases on every commit to main. 5000+ releases on GitHub :)

I think we should avoid that. Having too many releases will create notification fatigue and we'll lose watchers on the repo.

> This would leave us with only minor and major releases which we need to do manually via postgrest-release. Much better workflow, our users get their bugfixes and features much quicker.
> WDYT?

Agree overall :+1:

**Comment by laurenceisla** [comment]
@wolfgangwalther 

> Migrate the docs repo

Is it necessary to migrate the docs to older branches? Will they not stay as they are right now in RTD?

Otherwise, everything LGTM.

**Comment by steve-chavez** [comment]
> "Write docs for new features."

True. Those are missing.

**Comment by steve-chavez** [comment]
> I don't like keeping the old prereleases, though. This makes it basically impossible to use the release page as kind of a changelog to the project, because all the pre-releases keep repeating the same changes over and over again. Just have a look at page two: https://github.com/PostgREST/postgrest/releases?page=2

Yeah, don't like that too. Too much noise.

> @steve-chavez any objections?

> So we'd do a nightly or devel tag after each commit to main automatically. I like devel more, because we already use that as a version in the latest docs:

No objections.. `devel` looks good :+1:

**Comment by laurenceisla** [comment]
> Was there a specific reason or was it just not done/forgotten?

The latter.

> I would fix that once I come around to make the new GitHub Runner on ARM work nicely  

Awesome! 

> and then delete all the old -arm tags, too, given that the underlying images will still be there. WDYT?

Yeah, no need to keep those `-arm` tags. Go ahead!
